File: core/time/src/main/java/me/nalab/core/time/TimeUtil.java
package me.nalab.core.time;

import java.time.Clock;
import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;

public class TimeUtil {

	private static Clock clock = null;

	private TimeUtil() {
		throw new UnsupportedOperationException("Cannot invoke constructor \"TimeUtil()\"");
	}

	public static Instant toInstant() {
		var current = Instant.now();
		if (clock != null) {
			current = Instant.now(clock);
		}
		return formatTo6Digit(current);
	}

	private static Instant formatTo6Digit(Instant instant) {
		var formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSSSSSX")
			.withZone(ZoneId.of("UTC"));
		return Instant.parse(formatter.format(instant));
	}

	public static void fixed(Clock clock) {
		TimeUtil.clock = clock;
	}

	public static void clear() {
		TimeUtil.clock = null;
	}
}

File: src/main/kotlin/me/nalab/api/core/TimeUtil.kt
package me.nalab.api.core

import java.time.Clock
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter

object TimeUtil {

    private var clock: Clock? = null

    fun toInstant(): Instant {
        val current = if (clock != null) {
            Instant.now(clock)
        } else {
            Instant.now()
        }
        return formatTo6Digit(current)
    }

    private fun formatTo6Digit(instant: Instant): Instant {
        val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSSSSSX")
            .withZone(ZoneId.of("UTC"))
        return Instant.parse(formatter.format(instant))
    }

    fun fixed(clock: Clock?) {
        this.clock = clock
    }

    fun clear() {
        clock = null
    }
}

